src/Transaction.cc

```diff
@@ -0,0 +1,293 @@
+        ClientException::throwException(HERE, task->getStatus());
+    }
+
+    return (task->getDecision() == WireFormat::TxDecision::COMMIT);
+}
+
+/**
+ * Block until the decision of this transaction commit is accepted by all
+ * participant servers.  If the commit has not yet occurred and a decision is
+ * not yet reached, this method will also start the commit.
+ */
+void
+Transaction::sync()
+{
+    if (!commitStarted) {
+        commitStarted = true;
+        ramcloud->transactionManager.addTransactionTask(taskPtr);
+    }
+
+    ClientTransactionTask* task = taskPtr.get();
+
+    while (!task->isReady()) {
+        ramcloud->poll();
+    }
+
+    if (task->getStatus() != STATUS_OK) {
+        ClientException::throwException(HERE, task->getStatus());
+    }
+}
+
+/**
+ * Commits the transaction defined by the operations performed on this
+ * transaction (read, remove, write).  This method blocks until a participant
+ * servers have accepted the decision.
+ *
+ * \return
+ *      True if the transaction was able to commit.  False otherwise.
+ */
+bool
+Transaction::commitAndSync()
+{
+    sync();
+    ClientTransactionTask* task = taskPtr.get();
+    return (task->getDecision() == WireFormat::TxDecision::COMMIT);
+}
+
+/**
+ * Read the current contents of an object as part of this transaction.
+ *
+ * \param tableId
+ *      The table containing the desired object (return value from
+ *      a previous call to getTableId).
+ * \param key
+ *      Variable length key that uniquely identifies the object within tableId.
+ *      It does not necessarily have to be null terminated.
+ * \param keyLength
+ *      Size in bytes of the key.
+ * \param[out] value
+ *      After a successful return, this Buffer will hold the
+ *      contents of the desired object - only the value portion of the object.
+ */
+void
+Transaction::read(uint64_t tableId, const void* key, uint16_t keyLength,
+        Buffer* value)
+{
+    ReadOp readOp(this, tableId, key, keyLength, value);
+    readOp.wait();
+}
+
+/**
+ * Delete an object from a table as part of this transaction. If the object does
+ * not currently exist then the operation succeeds without doing anything.
+ *
+ * \param tableId
+ *      The table containing the object to be deleted (return value from
+ *      a previous call to getTableId).
+ * \param key
+ *      Variable length key that uniquely identifies the object within tableId.
+ *      It does not necessarily have to be null terminated.
+ * \param keyLength
+ *      Size in bytes of the key.
+ */
+void
+Transaction::remove(uint64_t tableId, const void* key, uint16_t keyLength)
+{
+    if (expect_false(commitStarted)) {
+        throw TxOpAfterCommit(HERE);
+    }
+
+    ClientTransactionTask* task = taskPtr.get();
+
+    Key keyObj(tableId, key, keyLength);
+    ClientTransactionTask::CacheEntry* entry = task->findCacheEntry(keyObj);
+
+    if (entry == NULL) {
+        entry = task->insertCacheEntry(tableId, key, keyLength, NULL, 0);
```

Comment:
Any reason why you don't reuse the existing keyObj in this API? Otherwise, insertCacheEntry just has to make a new one. Same in other methods below.